plugin/src/main/java/org/opensearch/ml/action/prediction/TransportPredictionTaskAction.java

```diff
@@ -137,9 +137,16 @@ private void executePredict(
         String requestId = mlPredictionTaskRequest.getRequestID();
         log.debug("receive predict request " + requestId + " for model " + mlPredictionTaskRequest.getModelId());
         long startTime = System.nanoTime();
+        // For remote text embedding model, neural search will set mlPredictionTaskRequest.getMlInput().getAlgorithm() as
+        // TEXT_EMBEDDING. In ml-commons we should always use the real function name of model: REMOTE. So we try to get
+        // from model cache first.
+        FunctionName functionName = modelCacheHelper
+            .getOptionalFunctionName(modelId)
+            .orElse(mlPredictionTaskRequest.getMlInput().getAlgorithm());
         mlPredictTaskRunner
             .run(
-                mlPredictionTaskRequest.getMlInput().getAlgorithm(),
+                // This is by design to NOT use mlPredictionTaskRequest.getMlInput().getAlgorithm() here
+                functionName,
```

Comment:
We should fine-tune this part either in neural search side or in ml-commons side, more and more twisted code could cause more and more issues.